app/models/cost_query/cache.rb

```diff
@@ -0,0 +1,75 @@
+#-- copyright
+# OpenProject Reporting Plugin
+#
+# Copyright (C) 2010 - 2014 the OpenProject Foundation (OPF)
+#
+# This program is free software; you can redistribute it and/or
+# modify it under the terms of the GNU General Public License
+# version 3.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program; if not, write to the Free Software
+# Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+#++
+
+module CostQuery::Cache
+  class << self
+
+    def check
+      reset! if reset_required?
+    end
+
+    def reset!
+      CostQuery::Filter.reset!
+      CostQuery::Filter::CustomFieldEntries.reset!
+      CostQuery::GroupBy.reset!
+      CostQuery::GroupBy::CustomFieldEntries.reset!
+
+      update_reset_on
+    end
+
+    protected
+
+    attr_accessor :custom_fields_updated_on,
+                  :custom_fields_id_sum
+
+    def invalid?
+      updated_on = fetch_custom_field_updated_at
+      id_sum = fetch_custom_fields_changed
+
+      custom_fields_updated_on != updated_on ||
+        custom_fields_id_sum != id_sum
+    end
+
+    def update_reset_on
+      return if caching_disabled?
+
+      self.custom_fields_updated_on = fetch_custom_field_updated_at
```

Comment:
Granted, this is brittle and might break in the future, but right now, you are not correct.

Because what happens is actually:

1. read max date from DB
2. clear all generated classes, write max updated on
3. Generate classes when needed

So the worst thing that could happen is that between 2 and 3 a custom field is updated/created. This would result in the `custom_fields_updated_on` being outdated. The classes are however generated when needed and are thus up to date. Of course, if the change is done after the classes are created we are out but that is IMO acceptable for the use case (filtering) we have. On the next request we would notice, that the stored classes are outdated (although they might in fact not be so) and recreate them. This is not great but good enough.

If you are referring to the the call to `update_reset_on` in https://github.com/finnlabs/openproject-reporting/pull/68/files#diff-9ad6fa05b3389aba75c319ddf2776d27R33 this will yield the same result as the first call (because of rails sql caching). However, I think moving it up to the top of the `reset!` method makes sense anyway.